### PR TITLE
grpc: Allow user defined grpc status

### LINKED ISSUE #8053: grpc_json_transcoder header not handled proply
`grpc-status` Header is missing when  `gRPC status` > Status::GrpcStatus::MaximumValid

https://github.com/envoyproxy/envoy/blob/master/source/extensions/filters/http/grpc_json_transcoder/json_transcoder_filter.cc#L430

Example:
```
grpcMethodA
    SetHeader("xyz", "123")
    return errorStatus(3), errorMessage("hello, error message")

< HTTP/2 400
< content-type: application/json
< xyz: 123
< x-envoy-upstream-service-time: 1
< grpc-status: 3
< grpc-message: hello, error message
< content-length: 0
< date: Tue, 27 Aug 2019 13:02:42 GMT
< server: istio-envoy

grpcMethodB
    SetHeader("xyz", "123")
    return errorStatus(1024), errorMessage("hello, error message")

< HTTP/2 503
< content-type: application/json
< xyz: 123
< x-envoy-upstream-service-time: 7
< grpc-message: hello, error message
< content-length: 0
< date: Tue, 27 Aug 2019 13:04:04 GMT
< server: istio-envoy
```

But when using Trailer, when StatusOK(0), trailer related header is missing, it only shows when error accurred.

### PR DESCRIPTION
**Pull request by Shikugawa**
Description: This fix is a branch of issue #8053. User defined grpc-status is allow on grpc specification so that it is truly handled on envoy grpc-json transcoding
Risk Level: Low
Testing: Unit Test
Docs Changes: none
Release Notes: none
[Optional Fixes #8053]
[Optional Deprecated:]

### CODE REVIEW

File: source/common/grpc/google_async_client_impl.cc [hunk 1]
```diff
@@ -84,7 +84,7 @@ GoogleAsyncClientImpl::GoogleAsyncClientImpl(Event::Dispatcher& dispatcher,
   stub_ = stub_factory.createStub(channel);
   // Initialize client stats.
   stats_.streams_total_ = &scope_->counter("streams_total");
-  for (uint32_t i = 0; i <= Status::GrpcStatus::MaximumValid; ++i) {
+  for (uint32_t i = 0; i <= Status::WellKnownGrpcStatus::UserDefinedGrpcStatus; ++i) {
```
Comment (lizan) [inline source/common/grpc/google_async_client_impl.cc]: I will probably keep this until MaximumValid, until we see some feature request for envoy grpc client. I don't think you increment this counter either?

File: include/envoy/grpc/status.h [hunk 2]
```diff
@@ -44,7 +46,10 @@ class Status {
     Unauthenticated = 16,
 
     // Maximum value of valid status codes.
-    MaximumValid = Unauthenticated,
+    MaximumKnown = Unauthenticated,
+
+    // user defined grpc status code.
+    UserDefinedGrpcStatus = MaximumKnown + 1,
```
Comment (lizan) [inline include/envoy/grpc/status.h]: I don't think you need this based on comment below.

File: source/common/grpc/common.cc [hunk 3]
```diff
@@ -56,8 +56,8 @@ absl::optional<Status::GrpcStatus> Common::getGrpcStatus(const Http::HeaderMap&
     return absl::nullopt;
   }
   if (!absl::SimpleAtoi(grpc_status_header->value().getStringView(), &grpc_status_code) ||
-      grpc_status_code > Status::GrpcStatus::MaximumValid) {
-    return {Status::GrpcStatus::InvalidCode};
+      grpc_status_code > Status::WellKnownGrpcStatus::MaximumKnown) {
```
Comment (lizan) [inline source/common/grpc/common.cc]: I think you want to allow the condition `grpc_status_code > Status::WellKnownGrpcStatus::MaximumKnown`? Though we might want to guard this with an additional parameter. Make this method `getGrpcStatus(const Http::HeaderMap& trailers, bool allow_user_defined = false)`?

Comment (Shikugawa) [reply source/common/grpc/common.cc]: I think if we applied `allow_user_defined` param, user must add management flag of this in grpc-json transcoder config. Is it ok?

Comment (Shikugawa) [reply source/common/grpc/common.cc]: I think if we applied allow_user_defined param, user must add management flag of this in grpc-json transcoder config. Is it ok?

File: include/envoy/grpc/status.h [hunk 4]
```diff
@@ -5,9 +5,11 @@ namespace Grpc {
 
 class Status {
 public:
+  using GrpcStatus = int64_t;
+
   // If this enum is changed, then the std::unordered_map in Envoy::Grpc::Utility::nameToGrpcStatus
   // located at: //source/common/access_log/grpc/status.cc must also be changed.
-  enum GrpcStatus {
+  enum WellKnownGrpcStatus {
```
Comment (yanavlasov) [inline include/envoy/grpc/status.h]: Renaming this enum type will cause build problems for all extensions that use it (and live outside of the envoy repository). Is it necessary that the uint64_t alias be named GrpcStatus? Could it be Type? Grpc::Status::Type may look better than the Grpc::Status::GrpcStatus

Comment (lizan) [reply include/envoy/grpc/status.h]: @yanavlasov we don't have C++ API stability. Renaming to Type also break the build outside of the tree too?

File: source/common/grpc/common.cc (hunk 3 above)
Comment (lizan) [reply source/common/grpc/common.cc]: In grpc-json-transcoder you should always set it to `true`, for other places keep it `false` should be fine.

File: include/envoy/grpc/status.h (hunk 4 above)
Comment (yanavlasov) [reply include/envoy/grpc/status.h]: I'm going to propose to have at least some common sense stability. We had to spend significant time fixing builds of internal filters for upstream C++ API changed that I think were totally avoidable.
Here I meant using Type in the int64_t alias and leaving enum GrpcStatus unchanged. I.e:
using Type = int64_t;
enum GrpcStatus { ... };

Comment (lizan) [reply include/envoy/grpc/status.h]: yeah if it is avoidable. If we use `Type`, don't you still have to change all `Grpc::Status::GrpcStatus` to `Grpc::Status::Type`?

Comment (yanavlasov) [reply include/envoy/grpc/status.h]: I see. Yes, you are correct, with this change filters have to change anyway. I looked at the internal filters and I think the original change with using GrpcStatus for int64_t alias is less problematic. So, please disregard my comment.

File: source/common/grpc/common.h [hunk 5]
```diff
@@ -45,10 +45,13 @@ class Common {
...
    * Returns the GrpcStatus code from a given set of trailers, if present.
    * @param trailers the trailers to parse.
+   * @param allow_user_status whether allow user defined grpc status.
+   *        if this value is false, custom grpc status is regarded as invalid status
    * @return absl::optional<Status::GrpcStatus> the parsed status code or InvalidCode if no valid
    * status is found.
    */
-  static absl::optional<Status::GrpcStatus> getGrpcStatus(const Http::HeaderMap& trailers);
+  static absl::optional<Status::GrpcStatus> getGrpcStatus(const Http::HeaderMap& trailers,
+                                                          bool allow_user_status = false);
```
Comment (lizan) [inline source/common/grpc/common.h]: nit: `allow_user_defined` (.cc is correct).

Comment (yanavlasov) [inline source/common/grpc/common.h]: I can't quite find if there are unit tests that check the new behavior (allow_user_status == true) of this method.

File: source/extensions/filters/http/grpc_json_transcoder/json_transcoder_filter.cc [hunk 6]
```diff
@@ -465,7 +465,7 @@ Http::FilterTrailersStatus JsonTranscoderFilter::encodeTrailers(Http::HeaderMap&
   response_in_.finish();
 
   const absl::optional<Grpc::Status::GrpcStatus> grpc_status =
-      Grpc::Common::getGrpcStatus(trailers);
+      Grpc::Common::getGrpcStatus(trailers, true);
```
Comment (yanavlasov) [inline source/extensions/filters/http/grpc_json_transcoder/json_transcoder_filter.cc]: Would it make sense to add tests to the json_transcoder_filter_test for this change?

**Comment by lizan** [comment]
@Shikugawa can you merge master and resolve conflicts?

File: source/common/config/grpc_mux_impl.cc [hunk 7]
```diff
@@ -0,0 +1,249 @@
+#include "common/config/grpc_mux_impl.h"
```
Comment (lizan) [inline source/common/config/grpc_mux_impl.cc]: this file shouldn't be included

**Pull request merged**